src/main/scala/com/fulcrumgenomics/bam/AddReadGroupsByName.scala

```diff
@@ -0,0 +1,182 @@
+  private val readGroups = new mutable.HashMap[RunInfo, SAMReadGroupRecord]()
+
+  override def execute(): Unit = {
+
+    // Gather all the read groups
+    {
+      val progress = new ProgressLogger(logger, verb = "read", unit = 5e6.toInt)
+      val in = SamReaderFactory.make().open(input.toFile)
+
+      in.iterator().foreach { record =>
+        val runInfo       = RunInfo(name=record.getReadName)
+        val platformModel = this.platformModel.getOrElse(runInfo.platformModel)
+
+        if (!readGroups.contains(runInfo)) {
+          val readGroup = new SAMReadGroupRecord(nextId.incrementAndGet().toString)
+
+          readGroup.setSample(sample)
+          readGroup.setLibrary(library)
+          readGroup.setPlatformUnit(runInfo.platformUnit)
+          readGroup.setPlatformModel(platformModel)
+          platform.foreach(readGroup.setPlatform)
+          sequencingCenter.foreach(readGroup.setSequencingCenter)
+          predictedInsertSize.foreach(readGroup.setPredictedMedianInsertSize)
+          description.foreach(readGroup.setDescription)
+          runDate.foreach(readGroup.setRunDate)
+
+          readGroups(runInfo) = readGroup
+        }
+        else {
+          // Check that the read information matches the read group
+          val readGroup = readGroups(runInfo)
+
+          // Used in the lookup of the read group, should never happen!
+          if (readGroup.getPlatformUnit != runInfo.platformUnit) unreachable(s"PU mismatch: s'${readGroup.getPlatformModel} != ${runInfo.platformUnit}")
+          if (readGroup.getPlatformModel != platformModel) unreachable(s"PM mismatch: s'${readGroup.getPlatformModel} != $platformModel")
+
+          // These should be set across all reads, so should never happen!
+          if (platform.exists(_ != readGroup.getPlatform)) unreachable(s"PL mismatch: s'${readGroup.getPlatform} != ${platform.get}")
+          if (sequencingCenter.exists(_ != readGroup.getSequencingCenter)) unreachable(s"CN mismatch: s'${readGroup.getSequencingCenter} != sequencingCenter.get}")
+          if (predictedInsertSize.exists(_ != readGroup.getPredictedMedianInsertSize)) unreachable(s"PI mismatch: s'${readGroup.getPredictedMedianInsertSize} != ${predictedInsertSize.get}")
+          if (description.exists(_ != readGroup.getDescription)) unreachable(s"DS mismatch: s'${readGroup.getDescription} != ${description.get}")
+          if (runDate.exists(_ != readGroup.getRunDate)) unreachable(s"DT mismatch: s'${readGroup.getRunDate} != ${runDate.get}")
+        }
+        progress.record(record)
+      }
+      in.safelyClose()
+    }
+
+    // Write them all out
+    {
+      val progress = new ProgressLogger(logger, verb = "written", unit = 5e6.toInt)
+      val in = SamReaderFactory.make().open(input.toFile)
+      val header = in.getFileHeader
+      val outHeader = header.clone()
+      outHeader.setSortOrder(sortOrder)
+      outHeader.setReadGroups(readGroups.values.toSeq)
```

Comment:
If you wanted to avoid using the implicit here I think you could also:
`Arrays.asList(readGroups.values.toSeq:_*)`